Fix fd race condition in multi-thread program

ReyRen opened the pull request:
The size argument is ignored since Linux 2.6.8 in "epoll_create". So,
We choose the "epoll_create1" and specific the close-on-exec flag FLAG
is "EPOLL_CLOEXEC". Please refer epoll_create(2).
Btw, the function "_fdset_cloexec" not stable than O_CLOEXEC flag to be
set directly, SOCK_CLOEXEC is used in socket fd. Because using a separate
fcntl(2) F_SETFD operation to set the FD_CLOEXEC flag does not suffice to
avoid race conditions where one thread opens a file descriptor and
attempts to set its close-on-exec flag using fcntl(2) at the same time as
another thread does a fork(2) plus execve(2).[referenced: open(2)]

fabbione commented:
test this please

fabbione commented:
a few things:

- can you please show me how to create this race condition?
- this code has not been tested at all, in fact it fails everywhere in CI. The passing of the flags |= with the socket looked weird.

Before submitting patches, please make sure to run make check and make check-memcheck (requires valgrind installed) locally.

yuanren10 committed:
Fix fd race condition in multi-thread program

The size argument is ignored since Linux 2.6.8 in "epoll_create". So,
We choose the "epoll_create1" and specific the close-on-exec flag FLAG
is "EPOLL_CLOEXEC". Please refer epoll_create(2).
Btw, the function "_fdset_cloexec" not stable than O_CLOEXEC flag to be
set directly, SOCK_CLOEXEC is used in socket fd. Because using a separate
fcntl(2) F_SETFD operation to set the FD_CLOEXEC flag does not suffice to
avoid race conditions where one thread opens a file descriptor and
attempts to set its close-on-exec flag using fcntl(2) at the same time as
another thread does a fork(2) plus execve(2).[referenced: open(2)]

ReyRen commented:
> a few things:
> 
> * can you please show me how to create this race condition?
> * this code has not been tested at all, in fact it fails everywhere in CI. The passing of the flags |= with the socket looked weird.
> 
> Before submitting patches, please make sure to run make check and make check-memcheck (requires valgrind installed) locally.

Actually, I didn't create rate condition yet, maybe I will. But the _fdset_cloexec and _fdset_nonblock not necessary to use because there are specific flags can be used and it seems like an atomic operation. (test successfully on local)

fabbione commented:
All the code in libknet is protected by the global write lock. There should be no race condition but i am happy to be proven wrong.

fabbione commented:
retest this please

fabbione commented:
The patch fails to build on FreeBSD because the EPOLL_CLOSEXEC changes are Linux specific and not portable.

ReyRen commented:
The epoll and EPOLL_CLOSEXEC used in "contrib which is Packages contributed by 3rd parties". So, it's not supported by original source tree. I will keep trying on the race condition, but for now, this or would be closed.

ReyRen commented:
btw, I didn't see any locked when you use fcntl to set fd flags after epoll create. I thought there might be interrupted by other thread between getting fd flags and set new flags. As for the FreeBSD, I have an idea is that once there is no HAVE_EPOLL_CLOEXEC defined, we just use zero as the parameter, because if someone has used contrib's library that would be all same with epoll_create1 on liunx. The reason I advice is that would clean some of our code and avoid race condition (still testing but it's a accidental event in multi-thread programs) maybe. What's your opinion about these, hoping for your advice. Thanks a lot.

fabbione commented:
the locks are at much higher level in the APIs. All public APIs are holding locks way before we get to call into fnctl.
